### ISSUE TITLE
Cannot cast from source type to destination type

### ISSUE THREAD

**Issue by drok** [drok/Harmony-CitiesSkylines#3]
```
ModException: The Mod C:\Program Files (x86)\Steam\steamapps\workshop\content\255710\2053500739 [AdvancedOutsideConnection.dll] has caused an error ---> System.InvalidCastException: Cannot cast from source type to destination type.
at Harmony.PatchInfoSerialization.Deserialize (byte[]) <0x000fc>
at Harmony.HarmonySharedState.GetPatchInfo (System.Reflection.MethodBase) <0x0006b>
at Harmony.PatchProcessor.Patch () <0x001bd>
at Harmony.HarmonyInstance.<PatchAll>b__9_0 (System.Type) <0x000ed>
at Harmony.CollectionExtensions.Do<System.Type> (System.Collections.Generic.IEnumerable`1<System.Type>,System.Action`1<System.Type>) <0x000a7>
at Harmony.HarmonyInstance.PatchAll (System.Reflection.Assembly) <0x00087>
at Harmony.HarmonyInstance.PatchAll () <0x000b8>
at AdvancedOutsideConnection.LoadingExtension.Init (bool) <0x0002e>
at AdvancedOutsideConnection.LoadingExtension.OnLevelLoaded (ICities.LoadMode) <0x00031>
at (wrapper dynamic-method) LoadingWrapper.LoadingWrapper.OnLevelLoaded_Patch5 (LoadingWrapper,SimulationManager/UpdateMode) <0x000b9>

  --- End of inner exception stack trace ---
```

**Comment by drok** [comment]
Adding [more detail](https://steamcommunity.com/workshop/filedetails/discussion/2399343344/4830646290525771073/#c4830646290527469054):

```
[quote=rkumob;4830646290527469054]FYI, I have https://steamcommunity.com/sharedfiles/filedetails/?id=2037888659 mod installed, so usually I just quit the game instantly. I've tried to exit to menu and got the game hang for half an hour (killed the process eventually). And I see a bunch of messages at the end of output.log like these:

[code]
PostProcessFX:Destroyed the old ConfigUI game object.
[SS] OnLevelUnloading
[SS] Saving speed: 0.65
Saving C:\Users\rkumo\AppData\Local\Colossal Order\Cities_Skylines\gameSettings.cgs  [Settings - Internal]
NetworkDetective | [Debug]  3,849,215ms | NetworkDetectiveTool.Remove()
InvalidCastException: Cannot cast from source type to destination type.
at Harmony.PatchInfoSerialization.Deserialize (byte[]) <0x000fc>
at Harmony.HarmonySharedState.GetPatchInfo (System.Reflection.MethodBase) <0x0006b>
at (wrapper dynamic-method) Harmony.PatchProcessor.Harmony.PatchProcessor.Unpatch_Patch0 (Harmony.PatchProcessor,Harmony.HarmonyPatchType,string) <0x000b2>
at Harmony.HarmonyInstance.Unpatch (System.Reflection.MethodBase,Harmony.HarmonyPatchType,string) <0x000c7>
at TrafficManager.LoadingExtension.RevertDetours () <0x0024f>
at TrafficManager.LoadingExtension.OnLevelUnloading () <0x0045f>
at LoadingWrapper.OnLevelUnloading () <0x00084>

Rethrow as ModException: The Mod C:\Program Files (x86)\Steam\steamapps\workshop\content\255710\1637663252 [CSUtil.CameraControl.dll, CSUtil.Commons.dll, TMPE.GenericGameBridge.dll, TMPE.API.dll, TMPE.CitiesGameBridge.dll, TrafficManager.dll] has caused an error

UnityEngine.DebugLogHandler:Internal_LogException(Exception, Object)
UnityEngine.DebugLogHandler:LogException(Exception, Object)
UnityEngine.Logger:LogException(Exception, Object)
UnityEngine.Debug:LogException(Exception)
LoadingWrapper:OnLevelUnloading()
LoadingManager:UnloadLevel()
PauseMenu:<Quit>m__2(UIComponent, Int32)
ColossalFramework.UI.UIView:PopModal(Int32)
ColossalFramework.UI.UIDynamicPanels:Hide(String, Int32)
ExitConfirmPanel:OnReturnMainMenu()
System.Object:DynamicEventWrapper_OnReturnMainMenu(ExitConfirmPanel, UIComponent, UIMouseEventParameter)
ColossalFramework.UI.UIComponent:OnClick(UIMouseEventParameter)
ColossalFramework.UI.UIButton:OnClick(UIMouseEventParameter)
ColossalFramework.UI.MouseHandler:ProcessInput(IInputTranslator, Ray, UIComponent, Boolean)
ColossalFramework.UI.UIInput:ProcessMouseInput()
ColossalFramework.UI.UIInput:FpsBoosterUpdate()
BehaviourUpdater.UiInputUpdater:UpdateInput()
BehaviourUpdater.Updater:Update()
 
(Filename:  Line: -1)

[PrecisionEngineering] OnLevelUnloading
[PrecisionEngineering] Manager Unload
PedBridgeTool.Remove()
Failed reverting a Harmony method patch Snooper.WorldInfoPanelPatches+WorldInfoPanel_UpdateBindings with ID 'com.cities_skylines.dymanoid.snooper'. Error message: System.InvalidCastException: Cannot cast from source type to destination type.
at Harmony.PatchInfoSerialization.Deserialize (byte[]) <0x000fc>
at Harmony.HarmonySharedState.GetPatchInfo (System.Reflection.MethodBase) <0x0006b>
at (wrapper dynamic-method) Harmony.PatchProcessor.Harmony.PatchProcessor.Unpatch_Patch0 (Harmony.PatchProcessor,Harmony.HarmonyPatchType,string) <0x000b2>
at Harmony.HarmonyInstance.Unpatch (System.Reflection.MethodBase,Harmony.HarmonyPatchType,string) <0x000c7>
at SkyTools.Patching.MethodPatcher/Patcher.RevertPatch (System.Reflection.MethodInfo) <0x00042>
at SkyTools.Patching.PatchBase.RevertPatch (SkyTools.Patching.IPatcher) <0x00063>
at SkyTools.Patching.MethodPatcher.Revert () <0x00093>

 
(Filename: C:/buildslave/unity/build/artifacts/generated/common/runtime/DebugBindings.gen.cpp Line: 51)

Failed reverting a Harmony method patch Snooper.HumanAIPatches+HumanAI_StartMoving1 with ID 'com.cities_skylines.dymanoid.snooper'. Error message: System.InvalidCastException: Cannot cast from source type to destination type.
at Harmony.PatchInfoSerialization.Deserialize (byte[]) <0x000fc>
at Harmony.HarmonySharedState.GetPatchInfo (System.Reflection.MethodBase) <0x0006b>
at (wrapper dynamic-method) Harmony.PatchProcessor.Harmony.PatchProcessor.Unpatch_Patch0 (Harmony.PatchProcessor,Harmony.HarmonyPatchType,string) <0x000b2>
at Harmony.HarmonyInstance.Unpatch (System.Reflection.MethodBase,Harmony.HarmonyPatchType,string) <0x000c7>
at SkyTools.Patching.MethodPatcher/Patcher.RevertPatch (System.Reflection.MethodInfo) <0x00042>
at SkyTools.Patching.PatchBase.RevertPatch (SkyTools.Patching.IPatcher) <0x00063>
at SkyTools.Patching.MethodPatcher.Revert () <0x00093>

[/code] [/quote]
```

[Interim report showing mods involved and load order](https://pastebin.com/Nf0jwz3P) ([alt](https://gist.github.com/drok/3b49adb7bc81b63e1d5b16ef8a219561)) provided by rkumob:


[2327536070  ZoomToCursor             ](https://steamcommunity.com/sharedfiles/filedetails/?id=2327536070)
[561699875                            ](https://steamcommunity.com/sharedfiles/filedetails/?id=561699875 )
[466158459   BuildingThemes           ](https://steamcommunity.com/sharedfiles/filedetails/?id=466158459 )
[557636025                            ](https://steamcommunity.com/sharedfiles/filedetails/?id=557636025 )
[530871278   DaylightClassic          ](https://steamcommunity.com/sharedfiles/filedetails/?id=530871278 )
[608517757   SubBuildingsTabBar       ](https://steamcommunity.com/sharedfiles/filedetails/?id=608517757 )
[1410003347  AdditiveShader           ](https://steamcommunity.com/sharedfiles/filedetails/?id=1410003347)
[2055465280  HealkitMod               ](https://steamcommunity.com/sharedfiles/filedetails/?id=2055465280)
[2053500739  AdvancedOutsideConnection](https://steamcommunity.com/sharedfiles/filedetails/?id=2053500739)
[2037888659  InstantReturnToDesktop   ](https://steamcommunity.com/sharedfiles/filedetails/?id=2037888659)
[1899640536  ThemeMixer               ](https://steamcommunity.com/sharedfiles/filedetails/?id=1899640536)
[1665106193  SkipIntro                ](https://steamcommunity.com/sharedfiles/filedetails/?id=1665106193)
[1612287735  PurchaseIt               ](https://steamcommunity.com/sharedfiles/filedetails/?id=1612287735)
[1591417160  HideIt                   ](https://steamcommunity.com/sharedfiles/filedetails/?id=1591417160)
[1209581656  LightingRebalance        ](https://steamcommunity.com/sharedfiles/filedetails/?id=1209581656)
[934994075   ServiceVehicleSelector2  ](https://steamcommunity.com/sharedfiles/filedetails/?id=934994075 )
[2175810604  Alaskan                  ](https://steamcommunity.com/sharedfiles/filedetails/?id=2175810604)
[2172488844  Picker                   ](https://steamcommunity.com/sharedfiles/filedetails/?id=2172488844)
[2170999920  HDRIH3Cubemap            ](https://steamcommunity.com/sharedfiles/filedetails/?id=2170999920)
[2126881996  ShadowDistanceFix        ](https://steamcommunity.com/sharedfiles/filedetails/?id=2126881996)
[2074258904  DemographicsMod          ](https://steamcommunity.com/sharedfiles/filedetails/?id=2074258904)
[2071210858  QuayUpgradeTool          ](https://steamcommunity.com/sharedfiles/filedetails/?id=2071210858)
[2051693155  BuildingVehicleManager   ](https://steamcommunity.com/sharedfiles/filedetails/?id=2051693155)
[2034713132  AutoRepair               ](https://steamcommunity.com/sharedfiles/filedetails/?id=2034713132)
[1925527132  AdvancedRoadTools        ](https://steamcommunity.com/sharedfiles/filedetails/?id=1925527132)
[1890830956  UndoMod                  ](https://steamcommunity.com/sharedfiles/filedetails/?id=1890830956)
[1844442251  FineRoadTool             ](https://steamcommunity.com/sharedfiles/filedetails/?id=1844442251)
[1844440354  FineRoadAnarchy          ](https://steamcommunity.com/sharedfiles/filedetails/?id=1844440354)
[1771841274  HDRIH2Cubemap            ](https://steamcommunity.com/sharedfiles/filedetails/?id=1771841274)
[1658679290  ForestBrush              ](https://steamcommunity.com/sharedfiles/filedetails/?id=1658679290)
[1625704117  RoundaboutBuilder        ](https://steamcommunity.com/sharedfiles/filedetails/?id=1625704117)
[1591584836  HDRIH1Cubemap            ](https://steamcommunity.com/sharedfiles/filedetails/?id=1591584836)
[1394468624  ImprovedStopSelection    ](https://steamcommunity.com/sharedfiles/filedetails/?id=1394468624)
[1388613752  TreeMovementControl      ](https://steamcommunity.com/sharedfiles/filedetails/?id=1388613752)
[1183931915  SkyboxReplacer           ](https://steamcommunity.com/sharedfiles/filedetails/?id=1183931915)
[928128676   ImprovedPublicTransport2 ](https://steamcommunity.com/sharedfiles/filedetails/?id=928128676 )
[898480258   CameraPositions          ](https://steamcommunity.com/sharedfiles/filedetails/?id=898480258 )
[881291183   PloppableAsphalt         ](https://steamcommunity.com/sharedfiles/filedetails/?id=881291183 )
[845665815   CSLMod.CSLMapView.Modding](https://steamcommunity.com/sharedfiles/filedetails/?id=845665815 )
[818641631   AmbientSoundsTuner2      ](https://steamcommunity.com/sharedfiles/filedetails/?id=818641631 )
[814498849   ImprovedContentManager   ](https://steamcommunity.com/sharedfiles/filedetails/?id=814498849 )
[812107110   LessSteam                ](https://steamcommunity.com/sharedfiles/filedetails/?id=812107110 )
[810858473   TrafficReport            ](https://steamcommunity.com/sharedfiles/filedetails/?id=810858473 )
[793176674   SharpTextures            ](https://steamcommunity.com/sharedfiles/filedetails/?id=793176674 )
[791221322   PropPrecision            ](https://steamcommunity.com/sharedfiles/filedetails/?id=791221322 )
[787611845   PropSnapping             ](https://steamcommunity.com/sharedfiles/filedetails/?id=787611845 )
[785528371   CinematicCameraExtended  ](https://steamcommunity.com/sharedfiles/filedetails/?id=785528371 )
[767233815   DecalFix                 ](https://steamcommunity.com/sharedfiles/filedetails/?id=767233815 )
[762520291   ShadowStrengthAdjuster   ](https://steamcommunity.com/sharedfiles/filedetails/?id=762520291 )
[694512541   PropLineTool             ](https://steamcommunity.com/sharedfiles/filedetails/?id=694512541 )
[689937287   SurfacePainter           ](https://steamcommunity.com/sharedfiles/filedetails/?id=689937287 )
[672248733   UltimateEyecandy         ](https://steamcommunity.com/sharedfiles/filedetails/?id=672248733 )
[667342976   LoadingScreenMod         ](https://steamcommunity.com/sharedfiles/filedetails/?id=667342976 )
[650436109   QuayAnarchy              ](https://steamcommunity.com/sharedfiles/filedetails/?id=650436109 )
[618536610   ASTKrasBaseSoundPack     ](https://steamcommunity.com/sharedfiles/filedetails/?id=618536610 )
[593588108   PropAnarchy              ](https://steamcommunity.com/sharedfiles/filedetails/?id=593588108 )
[591137516   RedcometMunicipalSoundPac](https://steamcommunity.com/sharedfiles/filedetails/?id=591137516 )
[523824395   FogOptions               ](https://steamcommunity.com/sharedfiles/filedetails/?id=523824395 )
[515489008   ElevatedTrainStationTrack](https://steamcommunity.com/sharedfiles/filedetails/?id=515489008 )
[502750307   NaturalResourcesBrush    ](https://steamcommunity.com/sharedfiles/filedetails/?id=502750307 )
[412146081   PostProcessFX            ](https://steamcommunity.com/sharedfiles/filedetails/?id=412146081 )
[406354745   SpeedSlider              ](https://steamcommunity.com/sharedfiles/filedetails/?id=406354745 )
[1548831935  AdvancedVehicleOptionsUID](https://steamcommunity.com/sharedfiles/filedetails/?id=1548831935)
[1415090282  AutoLineColor            ](https://steamcommunity.com/sharedfiles/filedetails/?id=1415090282)
[625608461   AdaptivePropVisibilityDis](https://steamcommunity.com/sharedfiles/filedetails/?id=625608461 )
[531738447   CSLShowMoreLimits        ](https://steamcommunity.com/sharedfiles/filedetails/?id=531738447 )
[2048266761  BuildingUsage            ](https://steamcommunity.com/sharedfiles/filedetails/?id=2048266761)
[2027161563  LifecycleRebalance       ](https://steamcommunity.com/sharedfiles/filedetails/?id=2027161563)
[2275749683  TrolleybusTrailerMod     ](https://steamcommunity.com/sharedfiles/filedetails/?id=2275749683)
[2133705267  ABLC                     ](https://steamcommunity.com/sharedfiles/filedetails/?id=2133705267)
[2187834843  FindIt                   ](https://steamcommunity.com/sharedfiles/filedetails/?id=2187834843)
[2019097300  HideUnconnectedTracks.Lif](https://steamcommunity.com/sharedfiles/filedetails/?id=2019097300)
[2380878816  TrainDisplay             ](https://steamcommunity.com/sharedfiles/filedetails/?id=2380878816)
[650805785   FPSCamera                ](https://steamcommunity.com/sharedfiles/filedetails/?id=650805785 )
[2383512572  BirdFlocksPOModule       ](https://steamcommunity.com/sharedfiles/filedetails/?id=2383512572)
[2376128727  RotateAroundModule       ](https://steamcommunity.com/sharedfiles/filedetails/?id=2376128727)
[2367577600  BackAndForthModule       ](https://steamcommunity.com/sharedfiles/filedetails/?id=2367577600)
[2367576372  SelfRotationModule       ](https://steamcommunity.com/sharedfiles/filedetails/?id=2367576372)
[1094334744  ProceduralObjects        ](https://steamcommunity.com/sharedfiles/filedetails/?id=1094334744)
[2101551127  Repaint                  ](https://steamcommunity.com/sharedfiles/filedetails/?id=2101551127)
[2139035212  GarbageBinManager        ](https://steamcommunity.com/sharedfiles/filedetails/?id=2139035212)
[2152013104  NetworkDetective         ](https://steamcommunity.com/sharedfiles/filedetails/?id=2152013104)
[2389228470  TransferBroker           ](https://steamcommunity.com/sharedfiles/filedetails/?id=2389228470)
[1637663252  TrafficManager           ](https://steamcommunity.com/sharedfiles/filedetails/?id=1637663252)
[2085403475  NodeController.LifeCycle ](https://steamcommunity.com/sharedfiles/filedetails/?id=2085403475)
[445589127   PrecisionEngineering     ](https://steamcommunity.com/sharedfiles/filedetails/?id=445589127 )
[1758376843  NetworkSkins             ](https://steamcommunity.com/sharedfiles/filedetails/?id=1758376843)
[1934023593  HideCrosswalks           ](https://steamcommunity.com/sharedfiles/filedetails/?id=1934023593)
[2030755273  PedestrianBridge         ](https://steamcommunity.com/sharedfiles/filedetails/?id=2030755273)
[2405917899  RON                      ](https://steamcommunity.com/sharedfiles/filedetails/?id=2405917899)
[2389414419  ZoningAdjuster           ](https://steamcommunity.com/sharedfiles/filedetails/?id=2389414419)
[781767563   IINS.ExtendedInfo        ](https://steamcommunity.com/sharedfiles/filedetails/?id=781767563 )
[465318661   ToggleableWhiteness      ](https://steamcommunity.com/sharedfiles/filedetails/?id=465318661 )
[1435741602  Snooper                  ](https://steamcommunity.com/sharedfiles/filedetails/?id=1435741602)
[426460372   FavoriteCims             ](https://steamcommunity.com/sharedfiles/filedetails/?id=426460372 )
[1556715327  ShowIt                   ](https://steamcommunity.com/sharedfiles/filedetails/?id=1556715327)
[1627986403  BulldozeIt               ](https://steamcommunity.com/sharedfiles/filedetails/?id=1627986403)
[1656549865  RebuildIt                ](https://steamcommunity.com/sharedfiles/filedetails/?id=1656549865)
[1741725093  NoteIt                   ](https://steamcommunity.com/sharedfiles/filedetails/?id=1741725093)
[1768810491  MeasureIt                ](https://steamcommunity.com/sharedfiles/filedetails/?id=1768810491)
[2346565561  SmarterFirefighters      ](https://steamcommunity.com/sharedfiles/filedetails/?id=2346565561)
[2349240408  AutoLineBudget           ](https://steamcommunity.com/sharedfiles/filedetails/?id=2349240408)
[2404099470  GravityModule            ](https://steamcommunity.com/sharedfiles/filedetails/?id=2404099470)
[1886877404  CustomEffectLoader       ](https://steamcommunity.com/sharedfiles/filedetails/?id=1886877404)
[1747800340                           ](https://steamcommunity.com/sharedfiles/filedetails/?id=1747800340)
[2367735356  UOCRevisited             ](https://steamcommunity.com/sharedfiles/filedetails/?id=2367735356)
[1764208250  MoreVehicles             ](https://steamcommunity.com/sharedfiles/filedetails/?id=1764208250)
[1721492498  CargoHoldFix             ](https://steamcommunity.com/sharedfiles/filedetails/?id=1721492498)
[1645781000  ResolveOverlaps          ](https://steamcommunity.com/sharedfiles/filedetails/?id=1645781000)
[2197863850  BOB                      ](https://steamcommunity.com/sharedfiles/filedetails/?id=2197863850)
[2325122732  SupplyChainColoring      ](https://steamcommunity.com/sharedfiles/filedetails/?id=2325122732)
[412149127   FontSelector             ](https://steamcommunity.com/sharedfiles/filedetails/?id=412149127 )
[498386331   BuildAnywhere            ](https://steamcommunity.com/sharedfiles/filedetails/?id=498386331 )
[2104976832  DirectConnectRoads.LifeCy](https://steamcommunity.com/sharedfiles/filedetails/?id=2104976832)
[2243670411  RealPop2                 ](https://steamcommunity.com/sharedfiles/filedetails/?id=2243670411)
[2131808622  PloppableRICO            ](https://steamcommunity.com/sharedfiles/filedetails/?id=2131808622)
[2255219025  UnifiedUI.LifeCycle      ](https://steamcommunity.com/sharedfiles/filedetails/?id=2255219025)
[2101977903  TransportVehicleReturnPat](https://steamcommunity.com/sharedfiles/filedetails/?id=2101977903)
[1619685021  MoveIt                   ](https://steamcommunity.com/sharedfiles/filedetails/?id=1619685021)
[1274199764  NetworkTiling            ](https://steamcommunity.com/sharedfiles/filedetails/?id=1274199764)
[2159934925  NodeMarkup               ](https://steamcommunity.com/sharedfiles/filedetails/?id=2159934925)
[2399343344  HarmonyMod               ](https://steamcommunity.com/sharedfiles/filedetails/?id=2399343344)

**Comment by drok** [comment]
Fixed in 8f1c006
[Verified by OR](https://steamcommunity.com/workshop/filedetails/discussion/2399343344/4830646290525771073/#c4830646290527323228)
